### PR TITLE
add testing PT 1.12

### LINKED ISSUE #13348: Enable PyTorch 1.12 testing
## 🚀 Feature

See title. This needs to be done before the next feature release.

The 1.12 release is planned for 06/28

### Motivation

Make sure latest and greatest works well
______________________________________________________________________

#### If you enjoy Lightning, check out our other projects! ⚡

- [**Metrics**](https://github.com/Lightning-AI/metrics): Machine learning metrics for distributed, scalable PyTorch applications.

- [**Lite**](https://pytorch-lightning.readthedocs.io/en/latest/starter/lightning_lite.html): enables pure PyTorch users to scale their existing code on any kind of device while retaining full control over their own loops and optimization logic.

- [**Flash**](https://github.com/Lightning-AI/lightning-flash): The fastest way to get a Lightning baseline! A collection of tasks for fast prototyping, baselining, fine-tuning, and solving problems with deep learning.

- [**Bolts**](https://github.com/Lightning-AI/lightning-bolts): Pretrained SOTA Deep Learning models, callbacks, and more for research and production with PyTorch Lightning and PyTorch.

- [**Lightning Transformers**](https://github.com/Lightning-AI/lightning-transformers): Flexible interface for high-performance research using SOTA Transformers leveraging Pytorch Lightning, Transformers, and Hydra.

cc @carmocca @akihironitta @borda

### PR DESCRIPTION
**Pull request by Borda**
## What does this PR do?

Resolves #13348

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the [CHANGELOG](https://github.com/Lightning-AI/lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

**Review by SeanNaren — APPROVED** [review]

File: .github/workflows/ci-pytorch_test-conda.yml [hunk 1]
```diff
@@ -24,11 +24,12 @@ jobs:
     strategy:
       fail-fast: false
       matrix:
-        python-version: ["3.8"]  # previous to last Python version as that one is already used in test-full
-        pytorch-version: ["1.9", "1.10"]
         # nightly: add when there's a release candidate
         include:
+          - {python-version: "3.7", pytorch-version: "1.9"}
```
Comment (rohitgr7) [inline .github/workflows/ci-pytorch_test-conda.yml:29]: why python version is changed for PT-1.9?

**Review by rohitgr7 — APPROVED** [review]

File: .github/workflows/ci-pytorch_test-conda.yml (hunk 1 above)
Comment (Borda) [reply .github/workflows/ci-pytorch_test-conda.yml:29]: just diversification, not a specific reason...
there is still a high number of users with py3.7

File: .github/workflows/ci-pytorch_test-conda.yml [hunk 2]
```diff
@@ -24,11 +24,12 @@ jobs:
...
+          - {python-version: "3.8", pytorch-version: "1.10"}
           - {python-version: "3.9", pytorch-version: "1.11"}
+          - {python-version: "3.9", pytorch-version: "1.12"}
```
[the remark this answers is not in the document]
Comment (Borda) [reply .github/workflows/ci-pytorch_test-conda.yml:32]: yep, but first need to get it running :)

**Comment by justusschock** [comment]
@Borda

> not sure why the 1.12 is not found :/

The dockerfile gives `https://download.pytorch.org/whl/cu${CUDA_VERSION_MM}/torch_stable.html` but the install instructions give `https://download.pytorch.org/whl/cu${CUDA_VERSION_MM}`

Also, so far 1.12 is not available in stablee, but only in nightlies until it is released.

**Comment by Borda** [comment]
> Also, so far 1.12 is not available in stablee, but only in nightlies until it is released.

are we both talking about Conda environment? because so far I know Conda uses channels but does not define search URLs...

**Review by otaj — APPROVED** [review]

**Review by carmocca — APPROVED** [review]

**Review by akihironitta — APPROVED** [review]

File: .actions/assistant.py [hunk 3]
```diff
@@ -42,6 +42,13 @@ def requirements_prune_pkgs(packages: Sequence[str], req_files: Sequence[str] =
...
 
+    @staticmethod
+    def _req_name(line: str) -> str:
+        idxs = [line.index(c) for c in "<=>,@" if c in line]
+        if idxs:
+            line = line[: min(idxs)]
+        return line.strip()
+
     @staticmethod
     def _prune_packages(req_file: str, packages: Sequence[str]) -> None:
```
[the remark this answers is not in the document]
Comment (Borda) [reply .actions/assistant.py:53]: :+1:

File: .github/workflows/ci-pytorch_test-conda.yml [hunk 4]
```diff
@@ -45,6 +46,10 @@ jobs:
         conda list
         pip install -e .[test]
 
+    - name: Freeze PIL (hotfix)
+      # import of PILLOW_VERSION which they recently removed in v9.0 in favor of __version__
```
[the remark this answers is not in the document]
Comment (Borda) [reply .github/workflows/ci-pytorch_test-conda.yml:51]: not really, tbh, not sure why it popped up now

**Comment by carmocca** [comment]
How are we doing here? What are the blockers? Any way I can help?

We need to finish this soonish to make sure native fsdp works well for the release

**Comment by Borda** [comment]
> We need to finish this soonish to make sure native fsdp works well for the release

THANK YOU for finishing it! :purple_heart: seems all is passing now, just dockers are still running... :rabbit:

File: .github/workflows/events-nightly.yml [hunk 5]
```diff
@@ -169,6 +169,7 @@ jobs:
           - {python_version: "3.8", pytorch_version: "1.9", cuda_version: "11.1"}
           - {python_version: "3.8", pytorch_version: "1.10", cuda_version: "11.1"}
           - {python_version: "3.9", pytorch_version: "1.11", cuda_version: "11.3.1"}
+          - {python_version: "3.9", pytorch_version: "1.12", cuda_version: "11.3.1"}
           #  nightly: add when there's a release candidate
           #  - {python_version: "3.9", pytorch_version: "1.12"}
```
Comment (carmocca) [inline .github/workflows/events-nightly.yml:174]: We can remove this

```suggestion
```

File: .github/workflows/ci-pytorch_test-full.yml [hunk 6]
```diff
@@ -100,7 +100,8 @@ jobs:
       run: |
         # adjust versions according installed Torch version
         python ./requirements/pytorch/adjust-versions.py requirements/pytorch/extra.txt
-        pip install --requirement ./requirements/pytorch/extra.txt --find-links https://download.pytorch.org/whl/cpu/torch_stable.html --upgrade
+        cat requirements/pytorch/extra.txt  # fixme
```
Comment (carmocca) [inline .github/workflows/ci-pytorch_test-full.yml:103]: Is this FIXME still relevant?

**Review by carmocca — APPROVED** [review]

**Review by akihironitta — APPROVED** [review]

File: .github/workflows/ci-pytorch_test-full.yml (hunk 6 above)
Comment (Borda) [reply .github/workflows/ci-pytorch_test-full.yml:103]:
```suggestion
```

File: .github/workflows/ci-pytorch_test-conda.yml (hunk 4 above)
[the remark this answers is not in the document]
Comment (Borda) [reply .github/workflows/ci-pytorch_test-conda.yml:71]: lets leave it for another PR :rabbit:

File: .github/workflows/ci-pytorch_test-conda.yml (hunk 2 above)
[the remark this answers is not in the document]
Comment (Borda) [reply .github/workflows/ci-pytorch_test-conda.yml:30]: ok, mind adding it to this PR? :)

[the remark this answers is not in the document]
Comment (Borda) [reply .github/workflows/ci-pytorch_test-conda.yml:30]: I made the edit but accidentally committed to wrong branch a this got merged before I got it to right one :(
